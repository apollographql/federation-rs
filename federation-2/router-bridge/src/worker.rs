use crate::error::Error;
use async_channel::{bounded, Receiver, Sender};
use deno_core::Op;
use deno_core::{op, Extension, OpState};
use serde::de::DeserializeOwned;
use serde::Deserialize;
use serde::Serialize;
use std::borrow::Cow;
use std::cell::RefCell;
use std::collections::hash_map::DefaultHasher;
use std::collections::HashMap;
use std::convert::TryFrom;
use std::fmt::Debug;
use std::hash::Hasher;
use std::rc::Rc;
use std::sync::Arc;
use std::thread::JoinHandle;
use tokio::sync::{oneshot, Mutex};

#[derive(Serialize, Deserialize, Debug)]
struct JsonPayload {
    id: String,
    payload: serde_json::Value,
}

pub(crate) struct JsWorker {
    response_senders: Arc<Mutex<HashMap<String, oneshot::Sender<serde_json::Value>>>>,
    response_receivers: Arc<Mutex<HashMap<String, oneshot::Receiver<serde_json::Value>>>>,
    sender: Sender<JsonPayload>,
    handle: Option<JoinHandle<()>>,
    unsent_plans: Arc<Mutex<HashMap<String, serde_json::Value>>>,
}

impl JsWorker {
    pub(crate) fn new(worker_source_code: &'static str) -> Self {
        let response_senders: Arc<Mutex<HashMap<String, oneshot::Sender<serde_json::Value>>>> =
            Default::default();

        let cloned_senders = response_senders.clone();

        let (response_sender, receiver) = bounded::<JsonPayload>(10_000);
        let (sender, request_receiver) = bounded::<JsonPayload>(10_000);

        let unsent_plans = Arc::new(Mutex::new(HashMap::new()));
        let my_unsent_plans = unsent_plans.clone();

        tokio::spawn(async move {
            while let Ok(json_payload) = receiver.recv().await {
                if let Some(sender) = cloned_senders.lock().await.remove(&json_payload.id) {
                    if let Err(e) = sender.send(json_payload.payload.clone()) {
                        // Keep our plan in our failed plan cache. Someone else might want it.
                        tracing::error!("jsworker: couldn't send json response: {:?}", e);
                        my_unsent_plans
                            .lock()
                            .await
                            .insert(json_payload.id, json_payload.payload);
                    }
                } else {
                    tracing::error!(
                        "jsworker: couldn't find sender for payload id {}",
                        &json_payload.id
                    );
                }
            }
            tracing::debug!("deno runtime shutdown successfully");
        });

        let handle = std::thread::spawn(move || {
            let my_ext = Extension {
                name: concat!(env!("CARGO_PKG_NAME"), "_worker"),
                ops: Cow::Borrowed(&[
                    send::DECL,
                    receive::DECL,
                    log_trace::DECL,
                    log_debug::DECL,
                    log_info::DECL,
                    log_warn::DECL,
                    log_error::DECL,
                ]),
                op_state_fn: Some(Box::new(move |state| {
                    state.put(response_sender.clone());
                    state.put(request_receiver);
                })),
                ..Default::default()
            };

            let mut js_runtime =
                crate::js::Js::new("query planner".to_string()).build_js_runtime(my_ext);

            let runtime = tokio::runtime::Builder::new_current_thread()
                .enable_all()
                .build()
                .unwrap();

            let future = async move {
                js_runtime
                    .execute_script_static("worker.js", worker_source_code)
                    .unwrap();
                js_runtime.run_event_loop(false).await
            };
            runtime.block_on(future).unwrap();
        });

        Self {
            sender,
            handle: Some(handle),
            response_receivers: Default::default(),
            response_senders,
            unsent_plans,
        }
    }

    pub(crate) async fn request<Request, Response>(
        &self,
        command: Request,
    ) -> Result<Response, Error>
    where
        Request: std::hash::Hash + Serialize + Send + Debug + 'static,
        Response: DeserializeOwned + Send + Debug + 'static,
    {
        // Let's see if we already have this query plan in our failed delivery cache
        let mut hasher = DefaultHasher::new();
        command.hash(&mut hasher);
        // JavaScript can't process 64 bit numbers, so convert our hash to a string...
        let id = hasher.finish().to_string();

        if let Some(payload) = self.unsent_plans.lock().await.remove(&id) {
            serde_json::from_value(payload).map_err(|e| Error::ParameterDeserialization {
                message: format!("deno: couldn't deserialize response : `{e:?}`"),
                id,
            })
        } else {
            self.send(Some(id.clone()), command)
                .await
                .map_err(|e| Error::DenoRuntime(format!("couldn't send request {e}")))?;
            self.receive(id)
                .await
                .map_err(|e| Error::DenoRuntime(format!("request: couldn't receive response {e}")))
        }
    }

    pub(crate) async fn send<Request>(
        &self,
        id_opt: Option<String>,
        request: Request,
    ) -> Result<String, Error>
    where
        Request: std::hash::Hash + Serialize + Send + Debug + 'static,
    {
        let id = match id_opt {
            Some(id) => id,
            None => {
                let mut hasher = DefaultHasher::new();
                request.hash(&mut hasher);
                // JavaScript can't process 64 bit numbers, so convert our hash to a string...
                hasher.finish().to_string()
            }
        };

        let (sender, receiver) = oneshot::channel();
        {
            self.response_senders
                .lock()
                .await
                .insert(id.clone(), sender);
            self.response_receivers
                .lock()
                .await
                .insert(id.clone(), receiver);
        }
        let json_payload = JsonPayload {
            id: id.clone(),
            payload: serde_json::to_value(request).map_err(|e| Error::ParameterSerialization {
                message: format!("deno: couldn't serialize request : `{e:?}`"),
                name: "request".to_string(),
            })?,
        };

        self.sender
            .send(json_payload)
            .await
            .map_err(|e| Error::DenoRuntime(format!("send: couldn't send request {e}")))?;
        Ok(id)
    }

    async fn receive<Response>(&self, id: String) -> Result<Response, Error>
    where
        Response: DeserializeOwned + Send + Debug + 'static,
    {
        let receiver = self
            .response_receivers
            .lock()
            .await
            .remove(&id)
            .expect("couldn't find id in response_receivers");
        let payload = receiver.await.map_err(|e| {
            Error::DenoRuntime(format!("request: couldn't receive response: {e:?}"))
        })?;

        serde_json::from_value(payload).map_err(|e| Error::ParameterDeserialization {
            message: format!("deno: couldn't deserialize response : `{e:?}`"),
            id,
        })
    }

    fn quit(&mut self) -> Result<(), Error> {
        if let Some(handle) = self.handle.take() {
            handle.join().map_err(|_| {
                Error::DenoRuntime("couldn't wait for JsRuntime to finish".to_string())
            })
        } else {
            Ok(())
        }
    }
}

impl Drop for JsWorker {
    fn drop(&mut self) {
        self.quit().unwrap_or_else(|e| eprintln!("{e}"));
    }
}

// Logging capabilities
#[op]
fn log_trace(_: &mut OpState, message: String) -> Result<(), anyhow::Error> {
    tracing::trace!("{message}");
    Ok(())
}

#[op]
fn log_debug(_: &mut OpState, message: String) -> Result<(), anyhow::Error> {
    tracing::debug!("{message}");
    Ok(())
}

#[op]
fn log_info(_: &mut OpState, message: String) -> Result<(), anyhow::Error> {
    tracing::info!("{message}");
    Ok(())
}

#[op]
fn log_warn(_: &mut OpState, message: String) -> Result<(), anyhow::Error> {
    tracing::warn!("{message}");
    Ok(())
}

#[op]
fn log_error(_: &mut OpState, message: String) -> Result<(), anyhow::Error> {
    tracing::error!("{message}");
    Ok(())
}

#[op]
async fn send(state: Rc<RefCell<OpState>>, payload: JsonPayload) -> Result<(), anyhow::Error> {
    let sender = {
        let state = state.borrow();
        // we're cloning here because we don't wanna keep the borrow across an await point
        state.borrow::<Sender<JsonPayload>>().clone()
    };

    sender
        .send(payload)
        .await
        .map_err(|e| anyhow::anyhow!("couldn't send response {e}"))
}

#[op]
async fn receive(state: Rc<RefCell<OpState>>) -> Result<JsonPayload, anyhow::Error> {
    let receiver = {
        let state = state.borrow();
        state.borrow::<Receiver<JsonPayload>>().clone()
    };

    receiver
        .recv()
        .await
        .map_err(|e| anyhow::anyhow!("op_receive: couldn't send response {e}"))
}

#[cfg(test)]
mod worker_tests {
    use super::JsWorker;
    use serde::{Deserialize, Serialize};

    #[tokio::test]
    #[tracing_test::traced_test]
    async fn logging_works() {
        let expected_present_logs = [
            "TRACE router_bridge::worker: this is a Trace level log",
            "DEBUG router_bridge::worker: this is a Debug level log",
            "INFO router_bridge::worker: this is an Info level log",
            "WARN router_bridge::worker: this is a Warn level log",
            "ERROR router_bridge::worker: this is an Error level log",
        ];
        run_logger().await;
        logs_assert(|lines: &[&str]| {
            for log in expected_present_logs {
                assert!(
                    lines.iter().any(|line| line.ends_with(log)),
                    "couldn't find log `{}` in the traced logs:\n{}",
                    log,
                    lines.join("\n")
                );
            }

            Ok(())
        });
    }

    async fn run_logger() {
        #[derive(Serialize, Deserialize, Debug, PartialEq, Eq, Hash)]
        enum Kind {
            Trace,
            Debug,
            Info,
            Warn,
            Error,
            Exit,
        }

        #[derive(Serialize, Deserialize, Debug, PartialEq, Eq, Hash)]
        struct Command {
            kind: Kind,
            message: Option<String>,
        }
        let worker = JsWorker::new(include_str!("../bundled/test_logger_worker.js"));

        let trace_succeeded: bool = worker
            .request(Command {
                kind: Kind::Trace,
                message: Some("this is a Trace level log".to_string()),
            })
            .await
            .unwrap();

        let debug_succeeded: bool = worker
            .request(Command {
                kind: Kind::Debug,
                message: Some("this is a Debug level log".to_string()),
            })
            .await
            .unwrap();

        let info_succeeded: bool = worker
            .request(Command {
                kind: Kind::Info,
                message: Some("this is an Info level log".to_string()),
            })
            .await
            .unwrap();

        let warn_succeeded: bool = worker
            .request(Command {
                kind: Kind::Warn,
                message: Some("this is a Warn level log".to_string()),
            })
            .await
            .unwrap();

        let error_succeeded: bool = worker
            .request(Command {
                kind: Kind::Error,
                message: Some("this is an Error level log".to_string()),
            })
            .await
            .unwrap();

        // let's shutdown the js worker before we run assertions,
        // to prevent a potential hang
        let shutdown_succeeded: bool = worker
            .request(Command {
                kind: Kind::Exit,
                message: None,
            })
            .await
            .unwrap();

        assert!(warn_succeeded, "couldn't send warn log command");
        assert!(info_succeeded, "couldn't send info log command");
        assert!(debug_succeeded, "couldn't send debug log command");
        assert!(trace_succeeded, "couldn't send trace log command");
        assert!(error_succeeded, "couldn't send error log command");
        assert!(shutdown_succeeded, "couldn't send shutdown command");
    }

    #[tokio::test]
    // This test ensures crypto.getRandomValues can be called.
    // the uuid dependency relies on it since v9.0
    async fn test_get_random_values() {
        let mut worker = JsWorker::new(include_str!("../bundled/test_get_random_values.js"));

        worker.quit().unwrap();
    }
}
