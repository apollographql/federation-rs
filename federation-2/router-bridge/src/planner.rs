/*!
 * Instantiate a QueryPlanner from a schema, and perform query planning
*/

use crate::worker::JsWorker;
use serde::{de::DeserializeOwned, Deserialize, Serialize};
use std::fmt::{Debug, Display, Formatter};
use std::sync::Arc;
use thiserror::Error;

// ------------------------------------

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
/// Options for the query plan
pub struct QueryPlanOptions {
    /// Use auto fragmentation
    pub auto_fragmentization: bool,
}

/// Default options for query planning
impl QueryPlanOptions {
    /// Default query plan options
    pub fn default() -> QueryPlanOptions {
        QueryPlanOptions {
            auto_fragmentization: false,
        }
    }
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
/// This is the context which provides
/// all the information to plan a query against a schema
pub struct OperationalContext {
    /// The graphQL schema
    pub schema: String,
    /// The graphQL query
    pub query: String,
    /// The operation name
    pub operation_name: String,
}

#[derive(Debug, Error, Serialize, Deserialize, PartialEq)]
/// Container for planning errors
pub struct BridgeErrors {
    /// The contained errors
    pub errors: Vec<BridgeError>,
}

impl Display for BridgeErrors {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        f.write_fmt(format_args!(
            "Planning errors: {}",
            self.errors
                .iter()
                .map(|e| e.to_string())
                .collect::<Vec<String>>()
                .join(", ")
        ))
    }
}

/// An error which occurred during JavaScript planning.
///
/// The shape of this error is meant to mimic that of the error created within
/// JavaScript, which is a [`GraphQLError`] from the [`graphql-js`] library.
///
/// [`graphql-js`]: https://npm.im/graphql
/// [`GraphQLError`]: https://github.com/graphql/graphql-js/blob/3869211/src/error/GraphQLError.js#L18-L75
#[derive(Debug, Error, Serialize, Deserialize, PartialEq)]
pub struct BridgeError {
    /// A human-readable description of the error that prevented planning.
    pub message: Option<String>,
    /// [`BridgeErrorExtensions`]
    #[serde(deserialize_with = "none_only_if_value_is_null_or_empty_object")]
    pub extensions: Option<BridgeErrorExtensions>,
}

/// `none_only_if_value_is_null_or_empty_object`
///
/// This function returns Ok(Some(T)) if a T can be deserialized,
///
/// Ok(None) if data contains Null or an empty object,
/// And fails otherwise, including if the key is missing.
fn none_only_if_value_is_null_or_empty_object<'de, D, T>(data: D) -> Result<Option<T>, D::Error>
where
    D: serde::de::Deserializer<'de>,
    T: serde::Deserialize<'de>,
{
    #[derive(Deserialize)]
    #[serde(untagged)]
    enum OptionOrValue<T> {
        Opt(Option<T>),
        Val(serde_json::value::Value),
    }

    let as_option_or_value: Result<OptionOrValue<T>, D::Error> =
        serde::Deserialize::deserialize(data);

    match as_option_or_value {
        Ok(OptionOrValue::Opt(t)) => Ok(t),
        Ok(OptionOrValue::Val(obj)) => {
            if let serde_json::value::Value::Object(o) = &obj {
                if o.is_empty() {
                    return Ok(None);
                }
            }

            Err(serde::de::Error::custom(format!(
                "invalid neither null nor empty object: found {:?}",
                obj,
            )))
        }
        Err(e) => Err(e),
    }
}

impl Display for BridgeError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        if let Some(msg) = &self.message {
            f.write_fmt(format_args!("{code}: {msg}", code = self.code(), msg = msg))
        } else {
            f.write_str(self.code())
        }
    }
}

#[derive(Debug, Serialize, Deserialize, PartialEq)]
/// Error codes
pub struct BridgeErrorExtensions {
    /// The error code
    pub code: String,
}

/// An error that was received during planning within JavaScript.
impl BridgeError {
    /// Retrieve the error code from an error received during planning.
    pub fn code(&self) -> &str {
        match self.extensions {
            Some(ref ext) => &*ext.code,
            None => "UNKNOWN",
        }
    }
}

// ------------------------------------

#[derive(Deserialize, Debug)]
/// The result of a router bridge invocation
pub struct BridgeResult<T> {
    /// The data if the query was successfully run
    pub data: Option<T>,
    /// The errors if the query failed
    pub errors: Option<Vec<BridgeError>>,
}

impl<T> BridgeResult<T>
where
    T: DeserializeOwned + Send + Debug + 'static,
{
    /// Turn a BridgeResult into an actual Result
    pub fn into_result(self) -> Result<T, Vec<BridgeError>> {
        if let Some(data) = self.data {
            Ok(data)
        } else {
            Err(self.errors.unwrap_or_else(|| {
                vec![BridgeError {
                    message: Some("an unknown error occured".to_string()),
                    extensions: None,
                }]
            }))
        }
    }
}

/// A Deno worker backed query Planner.

pub struct Planner<T>
where
    T: DeserializeOwned + Send + Debug + 'static,
{
    worker: Arc<JsWorker<PlanCmd, BridgeResult<T>>>,
}

impl<T> Debug for Planner<T>
where
    T: DeserializeOwned + Send + Debug + 'static,
{
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("Planner").finish()
    }
}

impl<T> Planner<T>
where
    T: DeserializeOwned + Send + Debug + 'static,
{
    /// Instantiate a `Planner` from a schema string
    pub async fn new(schema: String) -> Result<Self, Vec<BridgeError>> {
        let worker =
            JsWorker::<PlanCmd, BridgeResult<T>>::new(include_str!("../js-dist/plan_worker.js"));
        let worker_is_set_up = worker
            .request(PlanCmd::UpdateSchema { schema })
            .await
            .map_err(|e| {
                vec![BridgeError {
                    message: Some(e.to_string()),
                    extensions: None,
                }]
            });

        // Both cases below the mean schema update failed.
        // We need to pay attention here.
        // returning early will drop the worker, which will join the jsruntime thread.
        // however the event loop will run for ever. We need to let the worker know it needs to exit,
        // before we drop the worker
        match worker_is_set_up {
            Err(setup_error) => {
                let _ = worker.request(PlanCmd::Exit).await;
                return Err(setup_error);
            }
            Ok(setup) => {
                if let Some(errors) = setup.errors {
                    let _ = worker.request(PlanCmd::Exit).await;
                    return Err(errors);
                }
            }
        }

        let worker = Arc::new(worker);

        Ok(Self { worker })
    }

    /// Plan a query against an instantiated query planner
    pub async fn plan(
        &self,
        query: String,
        operation_name: Option<String>,
    ) -> Result<BridgeResult<T>, crate::error::Error> {
        self.worker
            .request(PlanCmd::Plan {
                query,
                operation_name,
            })
            .await
    }
}

impl<T> Drop for Planner<T>
where
    T: DeserializeOwned + Send + Debug + 'static,
{
    fn drop(&mut self) {
        // Send a PlanCmd::Exit signal
        let worker_clone = self.worker.clone();
        let _ = std::thread::spawn(|| {
            let runtime = tokio::runtime::Builder::new_current_thread()
                .build()
                .unwrap();

            let _ = runtime.block_on(async move { worker_clone.send(PlanCmd::Exit).await });
        })
        .join();
    }
}

#[derive(Serialize, Debug, Clone)]
#[serde(tag = "kind")]
enum PlanCmd {
    UpdateSchema {
        schema: String,
    },
    #[serde(rename_all = "camelCase")]
    Plan {
        query: String,
        operation_name: Option<String>,
    },
    Exit,
}

#[cfg(test)]
mod tests {
    use super::*;

    const QUERY: &str = include_str!("testdata/query.graphql");
    const SCHEMA: &str = include_str!("testdata/schema.graphql");

    #[tokio::test]
    async fn it_works() {
        let planner = Planner::<serde_json::Value>::new(SCHEMA.to_string())
            .await
            .unwrap();

        let data = planner
            .plan(QUERY.to_string(), None)
            .await
            .unwrap()
            .data
            .unwrap();

        insta::assert_snapshot!(serde_json::to_string_pretty(&data).unwrap());
    }

    #[tokio::test]
    // A series of queries that should fail graphql-js's validate function.  The federation
    // query planning logic automatically does some validation in order to do its duties.
    // Some, but not all, of that validation is also handled by the graphql-js validator.
    // However, we are trying to assert that we are testing graphql-js validation, not
    // Federation's query planner validation.  So we run a few validations which we do not
    // expect to every show up in Federation's query planner validation.
    // This one is for the NoFragmentCyclesRule in graphql/validate
    async fn invalid_graphql_validation_1_is_caught() {
        let errors= vec![BridgeError {
                message: Some("Cannot spread fragment \"thatUserFragment1\" within itself via \"thatUserFragment2\".".to_string()),
                extensions: None,
            }];

        assert_errors(
            errors,
            // These two fragments will spread themselves into a cycle, which is invalid per NoFragmentCyclesRule.
            "\
        fragment thatUserFragment1 on User {
            id
            ...thatUserFragment2
        }
        fragment thatUserFragment2 on User {
            id
            ...thatUserFragment1
        }
        query { me { id ...thatUserFragment1 } }"
                .to_string(),
            None,
        )
        .await;
    }

    #[tokio::test]
    // A series of queries that should fail graphql-js's validate function.  The federation
    // query planning logic automatically does some validation in order to do its duties.
    // Some, but not all, of that validation is also handled by the graphql-js validator.
    // However, we are trying to assert that we are testing graphql-js validation, not
    // Federation's query planner validation.  So we run a few validations which we do not
    // expect to every show up in Federation's query planner validation.
    // This one is for the ScalarLeafsRule in graphql/validate
    async fn invalid_graphql_validation_2_is_caught() {
        let errors = vec![BridgeError {
            message: Some(
                "Field \"id\" must not have a selection since type \"ID!\" has no subfields."
                    .to_string(),
            ),
            extensions: None,
        }];

        assert_errors(
            errors,
            // This Book resolver requires a selection set, per the schema.
            "{ me { id { absolutelyNotAcceptableLeaf } } }".to_string(),
            None,
        )
        .await;
    }

    #[tokio::test]
    // A series of queries that should fail graphql-js's validate function.  The federation
    // query planning logic automatically does some validation in order to do its duties.
    // Some, but not all, of that validation is also handled by the graphql-js validator.
    // However, we are trying to assert that we are testing graphql-js validation, not
    // Federation's query planner validation.  So we run a few validations which we do not
    // expect to every show up in Federation's query planner validation.
    // This one is for NoUnusedFragmentsRule in graphql/validate
    async fn invalid_graphql_validation_3_is_caught() {
        let errors = vec![BridgeError {
            message: Some("Fragment \"UnusedTestFragment\" is never used.".to_string()),
            extensions: None,
        }];

        assert_errors(
            errors,
            // This Book resolver requires a selection set, per the schema.
            "fragment UnusedTestFragment on User { id } query { me { id } }".to_string(),
            None,
        )
        .await;
    }

    #[tokio::test]
    async fn invalid_federation_validation_is_caught() {
        let errors = vec![BridgeError {
            message: Some(
                "Must provide operation name if query contains multiple operations.".to_string(),
            ),
            extensions: None,
        }];

        assert_errors(
            errors, // This requires passing an operation name (because there are multiple operations)
            // but we have not done that! Therefore, we expect a validation error from planning.
            "query Operation1 { me { id } } query Operation2 { me { id } }".to_string(),
            None,
        )
        .await;
    }

    #[tokio::test]
    async fn invalid_deserialization_doesnt_panic() {
        let planner = Planner::<serde_json::Number>::new(SCHEMA.to_string()).await;

        dbg!(&planner);
        assert!(planner.is_err());
    }

    #[tokio::test]
    async fn invalid_schema_is_caught() {
        let expected_errors = vec![BridgeError {
            message: Some("Syntax Error: Unexpected Name \"Garbage\".".to_string()),
            extensions: None,
        }];

        let actual_error = Planner::<serde_json::Value>::new("Garbage".to_string())
            .await
            .unwrap_err();

        assert_eq!(expected_errors, actual_error);
    }

    #[tokio::test]
    async fn syntactically_incorrect_query_is_caught() {
        let errors = vec![BridgeError {
            message: Some("Syntax Error: Unexpected Name \"Garbage\".".to_string()),
            extensions: None,
        }];

        assert_errors(errors, "Garbage".to_string(), None).await;
    }

    #[tokio::test]
    async fn query_missing_subfields() {
        let expected_error_message = r#"Field "reviews" of type "[Review]" must have a selection of subfields. Did you mean "reviews { ... }"?"#;

        let errors = vec![BridgeError {
            message: Some(expected_error_message.to_string()),
            extensions: None,
        }];

        assert_errors(
            errors,
            // This query contains reviews, which requires subfields
            "query ExampleQuery { me { id reviews } }".to_string(),
            None,
        )
        .await;
    }

    #[tokio::test]
    async fn query_field_that_doesnt_exist() {
        let expected_error_message = r#"Cannot query field "thisDoesntExist" on type "Query"."#;
        let errors = vec![BridgeError {
            message: Some(expected_error_message.to_string()),
            extensions: None,
        }];

        assert_errors(
            errors,
            // This query contains reviews, which requires subfields
            "query ExampleQuery { thisDoesntExist }".to_string(),
            None,
        )
        .await;
    }

    async fn assert_errors(
        expected_errors: Vec<BridgeError>,
        query: String,
        operation_name: Option<String>,
    ) {
        let planner = Planner::<serde_json::Value>::new(SCHEMA.to_string())
            .await
            .unwrap();

        let actual = planner.plan(query, operation_name).await.unwrap();

        assert_eq!(expected_errors, actual.errors.unwrap());
    }
}

#[cfg(test)]
mod planning_error {
    use crate::planner::{BridgeError, BridgeErrorExtensions};

    #[test]
    #[should_panic(
        expected = "Result::unwrap()` on an `Err` value: Error(\"missing field `extensions`\", line: 1, column: 2)"
    )]
    fn deserialize_empty_planning_error() {
        let raw = "{}";
        serde_json::from_str::<BridgeError>(raw).unwrap();
    }

    #[test]
    #[should_panic(
        expected = "Result::unwrap()` on an `Err` value: Error(\"missing field `extensions`\", line: 1, column: 44)"
    )]
    fn deserialize_planning_error_missing_extension() {
        let raw = r#"{ "message": "something terrible happened" }"#;
        serde_json::from_str::<BridgeError>(raw).unwrap();
    }

    #[test]
    fn deserialize_planning_error_with_extension() {
        let raw = r#"{
            "message": "something terrible happened",
            "extensions": {
                "code": "E_TEST_CASE"
            }
        }"#;

        let expected = BridgeError {
            message: Some("something terrible happened".to_string()),
            extensions: Some(BridgeErrorExtensions {
                code: "E_TEST_CASE".to_string(),
            }),
        };

        assert_eq!(expected, serde_json::from_str(raw).unwrap());
    }

    #[test]
    fn deserialize_planning_error_with_empty_object_extension() {
        let raw = r#"{
            "extensions": {}
        }"#;
        let expected = BridgeError {
            message: None,
            extensions: None,
        };

        assert_eq!(expected, serde_json::from_str(raw).unwrap());
    }

    #[test]
    fn deserialize_planning_error_with_null_extension() {
        let raw = r#"{
            "extensions": null
        }"#;
        let expected = BridgeError {
            message: None,
            extensions: None,
        };

        assert_eq!(expected, serde_json::from_str(raw).unwrap());
    }
}
