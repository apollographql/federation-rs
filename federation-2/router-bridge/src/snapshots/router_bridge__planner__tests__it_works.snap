---
source: router-bridge/src/planner.rs
assertion_line: 127
expression: "serde_json::to_string_pretty(&data).unwrap()"
---
{
  "kind": "QueryPlan",
  "node": {
    "kind": "Fetch",
    "serviceName": "accounts",
    "variableUsages": [],
    "operation": "{me{name{first last}}}",
    "operationKind": "query"
  }
}
